### ISSUE TITLE
log to include filenames

### ISSUE THREAD

**Issue by epa** [genehack/Git-Wrapper#67]
It would be handy for log entries to have a `filenames` method which returns a list of files changed.  This would involve changing the `git log` call to a different format such as `git log --stat` and parsing it accordingly.

**Comment by genehack** [comment]
Isn't this information already available in the `modifications` attribute of the object returned when you call the `->log()` method?

**Comment by genehack** [comment]
Oh, wait, I think I just found a bug -- this information _should_ be available in `modifications`, but that isn't getting set up correctly. 

Could you check this section of the code -- https://github.com/genehack/Git-Wrapper/blob/master/lib/Git/Wrapper.pm#L219 -- and see if what's being set up there would meet your needs? Basically, each `Git::Wrapper::Log` object should have an arrayref of `Git::Wrapper::File::RawModification` objects associated with it; each of those `RawModification` objects has a 'filename' method. 

I'll do a test release tonight that fixes the bug I just found.

**Comment by epa** [comment]
That sounds great, thank you.  Let me know when the test release is out.

**Comment by genehack** [comment]
Okay, the test release is at https://metacpan.org/release/GENEHACK/Git-Wrapper-0.044_90 (or will be Real Soon once the index catches up).

If you have a look at a2077b243dbe24f5f0010e9041adcdcb15dcdd7e, you'll see a 'Raw logs' section I added which has a small demo program that illustrates how to pull out changed files on a per-commit basis, which is what I think you're after. Lemme know how that works for you.

**Comment by epa** [comment]
Thanks.  I think metacpan is still behind.  But why is the optional flag called `raw`?  Wouldn't something like `extra_info` be better?

**Comment by genehack** [comment]
`Git::Wrapper` methods just pass their arguments onto the underlying `git` command -- so it's `$git->log({ raw => 1 })` because the code is actually calling `git log --raw` and then parsing the results of that. 

A flag like `extra_info => 1` would violate this "arguments are command-line flags" API.

**Comment by epa** [comment]
Makes sense.  Is there a way I can download the test release given it's still not shown on metacpan?

**Comment by genehack** [comment]
It's on CPAN for me - https://metacpan.org/release/GENEHACK/Git-Wrapper-0.044_90-TRIAL

**Comment by epa** [comment]
Thanks, let me try it out.

**Comment by epa** [comment]
Looks good, thank you!  What I really would like to do is find all revisions that modify a particular file - but I don't know if git's command line interface has a way to do that directly.

**Comment by genehack** [comment]
`git log` can take a filename argument, and will then return only commits that involve that file.

**Comment by epa** [comment]
Hmm, yes of course it does.  I'm sure there was some reason why that didn't work and I ended up parsing the full log to look for the filename - but I can no longer remember what the reason was :-p.  Thanks for implementing the new feature.

**Comment by epa** [comment]
Thanks for implementing this.  But it doesn't correctly parse all git 'raw' logs.  For example

````
commit 9e5c1279c078a083d597dc9f0a77358086f646a5
Author: Ed Avis <eda@waniasset.com>
Date:   Tue Nov 25 14:52:07 2014 +0000

    Added Alpha 4 for Energy.


    git-svn-id: file:///home/svn/repos@39343 b388e47b-2472-4c64-b585-07cded36b160

:100644 100644 0456270ac. fcbfb7f... M  wwwroot/WaniSystems/Configuration/MarketSampler/jobs/build.trend.signals.xml
````

The line with `100644` doesn't match the regexp `m/^\:(\d{6}) (\d{6}) (\w{7})\.\.\. (\w{7})\.\.\. (\w{1})\t(.*)$/` and so the rest of the log ends up being skipped.  (I don't quite see how the rest is skipped, since there is a check at the top of the loop for unhandled lines, but it happens.)

I think there are a couple of things to fix: the regexp of course, but also tighten up the parsing so that a

**Comment by epa** [comment]
I suspect you may need `git log --raw --no-abbrev`.  The git manual page lies when it says that `raw` never abbreviates the SHA1s

**Comment by epa** [comment]
Git bug report: http://thread.gmane.org/gmane.comp.version-control.git/269148

**Comment by epa** [comment]
Apparently `--raw` is not it; you need `--format=raw`.  I haven't yet looked at how to parse that format.

**Comment by genehack** [comment]
`git log --format=raw` doesn't include a filename listing at all. 

It looks like adding `--no-abbrev` might work, but the parsing regexp will need updating. Give me a couple days for that (or beat me to it... 8^p )

**Comment by epa** [comment]
Yes I got thoroughly confused about the difference between 'raw format' and 'raw diff format' (the names of the command line options do not help at all - but that's git for you).  With `--no-abbrev` I believe you just need to change the regexp line to

````
      while(@out and $out[0] =~ m/^\:(\d{6}) (\d{6}) (\w{40}) (\w{40}) (\w{1})\t(.*)$/) {
````

The harder thing is how that flag should be passed down to git.  Should Git::Wrapper just add `--no-abbrev` whenever `raw` is specified?

**Comment by genehack** [comment]
It looks like (at least from a quick test) that unconditionally setting `$opt{no_abbrev} = 1` shouldn't have any other side-effects. There's probably some old-ass Git version that will throw an error because that option isn't supported, but the easiest way to find that is to have the code and just smoke it against a bunch of Git versions (sadly).

**Comment by genehack** [comment]
Okay, changes released and will show up at https://metacpan.org/release/GENEHACK/Git-Wrapper-0.044_91

Assuming there's no black smoke in the next few days, will do a full release.

**Comment by epa** [comment]
Thanks.  For version number of trial releases, could you use the _ rather than the - character?  This makes building packages easier since by convention - separates the version from the 'release' (if there are multiple package releases for the same upstream version).

**Comment by genehack** [comment]
Released 0.045.

(I'll see what I can do about future releases -- the '-TRIAL' bit getting added by the release tooling, I need to figure out how to turn that off.)
